src/order/conditionally_complete_lattice.lean

```diff
@@ -692,3 +716,137 @@ noncomputable instance with_top.with_bot.complete_lattice {α : Type*}
+the `Inf` of all its nonempty bounded-below subsets. -/
+noncomputable def subset_conditionally_complete_linear_order [inhabited s]
+  (h_Sup : ∀ {t : set s} (ht : t.nonempty) (h_bdd : bdd_above t), Sup (coe '' t : set α) ∈ s)
+  (h_Inf : ∀ {t : set s} (ht : t.nonempty) (h_bdd : bdd_below t), Inf (coe '' t : set α) ∈ s) :
+  conditionally_complete_linear_order s :=
+{ le_cSup := begin
+    rintros t c h_bdd hct,
+    -- The following would be a more natural way to finish, but gives a "deep recursion" error:
+    -- simpa [subset_Sup_of_within (h_Sup t)] using (strict_mono_coe s).monotone.le_cSup_image hct h_bdd,
+    have := (strict_mono_coe s).monotone.le_cSup_image hct h_bdd,
+    rwa subset_Sup_of_within s (h_Sup ⟨c, hct⟩ h_bdd) at this,
+  end,
+  cSup_le := begin
+    rintros t B ht hB,
+    have := (strict_mono_coe s).monotone.cSup_image_le ht hB,
+    rwa subset_Sup_of_within s (h_Sup ht ⟨B, hB⟩) at this,
+  end,
+  le_cInf := begin
+    intros t B ht hB,
+    have := (strict_mono_coe s).monotone.le_cInf_image ht hB,
+    rwa subset_Inf_of_within s (h_Inf ht ⟨B, hB⟩) at this,
+  end,
+  cInf_le := begin
+    rintros t c h_bdd hct,
+    have := (strict_mono_coe s).monotone.cInf_image_le hct h_bdd,
+    rwa subset_Inf_of_within s (h_Inf ⟨c, hct⟩ h_bdd) at this,
+  end,
+  ..subset_has_Sup s,
+  ..subset_has_Inf s,
+  ..distrib_lattice.to_lattice s,
+  ..classical.DLO s }
+
+section Ioo
+variables {a b : α}
+
+/-- The `Sup` function on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order takes values within `Ioo a b`, for all nonempty bounded-above sets. -/
+lemma Ioo_Sup_within {a b : α} ⦃t : set (Ioo a b)⦄ (ht : t.nonempty) (h_bdd : bdd_above t) :
+  Sup (coe '' t : set α) ∈ Ioo a b :=
+begin
+  obtain ⟨c, hct⟩ : ∃ c, c ∈ t := ht,
+  obtain ⟨B, hB⟩ : ∃ B, B ∈ upper_bounds t := h_bdd,
+  have ha := lt_of_lt_of_le c.2.1 ((strict_mono_coe (Ioo a b)).monotone.le_cSup_image hct ⟨B, hB⟩),
+  have hb := lt_of_le_of_lt ((strict_mono_coe (Ioo a b)).monotone.cSup_image_le ⟨c, hct⟩ hB) B.2.2,
+  exact ⟨ha, hb⟩,
+end
+
+/-- The `Inf` function on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order takes values within `Ioo a b`, for all nonempty bounded-above sets. -/
+lemma Ioo_Inf_within {a b : α} ⦃t : set (Ioo a b)⦄ (ht : t.nonempty) (h_bdd : bdd_below t) :
+  Inf (coe '' t : set α) ∈ Ioo a b :=
+begin
+  obtain ⟨c, hct⟩ : ∃ c, c ∈ t := ht,
+  obtain ⟨B, hB⟩ : ∃ B, B ∈ lower_bounds t := h_bdd,
+  have ha := lt_of_lt_of_le B.2.1 ((strict_mono_coe (Ioo a b)).monotone.le_cInf_image ⟨c, hct⟩ hB),
+  have hb := lt_of_le_of_lt ((strict_mono_coe (Ioo a b)).monotone.cInf_image_le hct ⟨B, hB⟩) c.2.2,
+  exact ⟨ha, hb⟩,
+end
+
+/-- A nonempty open interval of a conditionally complete linear order is naturally a conditionally
+complete linear order. -/
+noncomputable instance Ioo_conditionally_complete_linear_order [inhabited (Ioo a b)] :=
```

Comment:
If you have registered `ord_connected` as a typeclass, then this instance could be formulated for all `ord_connected` sets at once.